### PR TITLE
WIP: LeNet example

### LINKED ISSUE #2: Example of using convolutional bricks
I think we are really missing an example for those here. 

@harmdevries89, @dmitriy-serdyuk, do you have anything ready for that purpose?

### PR DESCRIPTION
**Pull request by dmitriy-serdyuk**
Simple convnet for MNIST with dropout.

Fixes #2.

### CODE REVIEW

File: tests/test_mnist_lenet.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+from six.moves import cPickle
+
+from mnist_lenet import main
+
+
+def test_mnist_lenet():
+    with tempfile.NamedTemporaryFile() as f:
+        main(f.name, 1, True)
+        with open(f.name, "rb") as source:
+            main_loop = cPickle.load(source)
```
Comment (bartvm) [inline tests/test_mnist_lenet.py]: Main loops serialized with the custom `dump` function can only be loaded with the custom `load` function.

File: mnist_lenet/__init__.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                        help=("Destination to save the state of the training "
+                              "process."))
+    parser.add_argument("--bokeh", action='store_true',
+                        help="Set if you want to use Bokeh ")
+    parser.add_argument("--feature-maps", type=int, default=None)
+    parser.add_argument("--mlp-hiddens", type=int, default=None)
+    parser.add_argument("--conv-sizes", type=int, default=None)
+    parser.add_argument("--pool-sizes", type=int, default=None)
+    args = parser.parse_args()
+    main(args.save_to, args.num_epochs, args.bokeh)
```
Comment (bartvm) [inline mnist_lenet/__init__.py]: You have a lot more arguments, which are never passed to `main` here.

File: mnist_lenet/__init__.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+from fuel.schemes import SequentialScheme
+from blocks.filter import VariableFilter
+from blocks.graph import ComputationGraph, apply_dropout
+from blocks.model import Model
+from blocks.monitoring import aggregation
+from blocks.extensions import FinishAfter, Timing, Printing
+from blocks.extensions.saveload import Checkpoint
+from blocks.extensions.monitoring import (DataStreamMonitoring,
+                                          TrainingDataMonitoring)
+from blocks.extensions.plot import Plot
```
Comment (bartvm) [inline mnist_lenet/__init__.py]: Sorry, I just moved the `Plot` extension to `blocks.extras`, but that's not being installed yet on Travis. Maybe it's easiest to get rid of the plotting for now. I'll add it back in a bit.

**Comment by bartvm** [comment]
Great work. I think it's worth keeping in mind https://github.com/mila-udem/blocks-examples/issues/5 i.e. add some explaining comments in the code here and there.

**Comment by dmitriy-serdyuk** [comment]
Yes, I'm going to add more comments and docstrings. I'm also using an old version of fuel, I'll fix it.

**Comment by dmitriy-serdyuk** [comment]
The PR will fix #2.

**Comment by dmitriy-serdyuk** [comment]
Do you know what should be the performance of LeNet with dropout? The current implementation gives around 1.18% error rate.

**Comment by bartvm** [comment]
No idea... Just had a look at the original paper, and it says they get 0.95% without dropout, so 1.18% seems high. They do use an Hessian approximation when training though. They also mention a parameter *μ*, but I can't figure out what that is.

I think it would be best instructive to replicate the model from the paper as closely as possible though i.e. without dropout.

**Comment by dmitriy-serdyuk** [comment]
I tuned hyperparameters a bit, now it reaches around 0.8% on the test set.

File: mnist_lenet/__init__.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        *convnet.children[0].get_dim('input_')))
+    for i, layer in enumerate(convnet.layers):
+        logging.info("Layer {} dim: {} {} {}".format(
+            i, *layer.get_dim('output')))
+
+    x = tensor.tensor4('features')
+    y = tensor.lmatrix('targets')
+
+    # Normalize input and apply the convnet
+    probs = convnet.apply((x / 256.) - 0.5)
```
Comment (lamblin) [inline mnist_lenet/__init__.py]: If `x` is in [0, 255], I think we should divide by `255.` instead.
Also, is it better to put that kind of normalization as a Transformer in data stream, or in the model directly? Is there a way we would like to encourage more?

Comment (dmitriy-serdyuk) [inline mnist_lenet/__init__.py]: Right, I guess the right way to it is in a transformer.

**Comment by dmitriy-serdyuk** [comment]
Cool!

**Comment by rizar** [comment]
Seems this nice demo is ready, should we merge it?

**Comment by dmitriy-serdyuk** [comment]
Yes, it is ready.

**Comment by rizar** [comment]
Awesome, great job!
